zerver/lib/markdown/__init__.py

```diff
@@ -574,14 +574,17 @@ def get_tweet_id(url: str) -> Optional[str]:
     return tweet_id_match.group("tweetid")
 
 
-class InlineHttpsProcessor(markdown.treeprocessors.Treeprocessor):
+class InlineImageProcessor(markdown.treeprocessors.Treeprocessor):
     def run(self, root: Element) -> None:
         # Get all URLs from the blob
         found_imgs = walk_tree(root, lambda e: e if e.tag == "img" else None)
         for img in found_imgs:
             url = img.get("src")
             assert url is not None
-            if urllib.parse.urlsplit(url).scheme != "http":
+            split_url = urllib.parse.urlsplit(url)
+            if (split_url.scheme == "https" and split_url.netloc in settings.ALLOWED_HOSTS) or (
```

Comment:
This check won't work -- `ALLOWED_HOSTS` can contain wildcards.  

We do need to exclude user uploads -- there's some imperfect functions for that question in `zerver/lib/thumbnail.py` that can at least help constrain the search space.  But structurally I think what we want to know is whether (1) the URL is the current organization's `realm.uri`, which we have access to:
```
        _md_engine.zulip_db_data = {                                                                 
            "realm_alert_words_automaton": realm_alert_words_automaton,                              
            "mention_data": mention_data,                                                            
            "active_realm_emoji": active_realm_emoji,                                                
            "realm_uri": message_realm.uri,                                                          
            "sent_by_bot": sent_by_bot,                                                              
            "stream_names": stream_name_info,                                                        
            "translate_emoticons": translate_emoticons,                                              
        }                                                                                            
```

(or a local URL), and maybe (2) whether the URL is in the `/user_uploads/` namespace where authentication is required.  Though I think we could probably choose to skip (2) for this purpose -- we need to definitely not use camo for `/user_uploads` URLs as doing so would break authentication, but using camo for other URLs on the Zulip server's domain is a somewhat free choice.  Probably makes sense to mostly think about what's right for security and also subtle features that also involve uploaded files like custom emoji and avatars, since in practice I don't think images included in the Zulip webapp itself are likely to get a posted often.

Reply:
Good point about `ALLOWED_HOSTS`, I think @andersk's solution fixes that.

I'm not sure I agree about:

>  But I think it's also the wrong question -- we don't care what other realms might be hosted on this server on different subdomains -- we only care about the realm hosting the current user's organization.

Given that the reason to use camo is privacy, I don' think there's a reason to use camo for a request to `foo.zulipchat.com/whatever.png` that came from `bar.zulipchat.com` — there's no privacy benefit to proxying that, since the requests will all end up in the same logs anyways. And as you point out, it likely makes thinking about authentication significantly more difficult. If we care about realm here, we need to think about excluding things for authentication/authorization purposes, but if we just exclude images regardless of realm, we get that for free.

It seems like it makes things significantly easier to reason about to just exclude every image hosted on the server regardless of realm, and doing so doesn't seem to have any privacy disadvantages.

Reply:
I could imagine a benefit around cases where the set of Zulip servers hosted on a given installation is dynamic over time (we certainly have had organizations migrate to and from Zulip Cloud to self hosting, in some cases without changing their domain).

But I think the broader argument on my mind is that we want to avoid features where Zulip's behavior depends on what other organizations are hosted on the same Zulip installation, and the behavior implemented here would be one of our first violations of such an invariant.  (The one intentional one we have is letting one copy settings during account creation).

I think the cleaner security policy is to only look at `realm.uri` for that reason, and have unnecessary camo usage for the rare links to other Zulip organizations hosted by the same server.  But I'm still not sure I can think of a realistic situation where the decision would matter one way or the other.  (This change might make it hard for us to keep it a secret that `chat.secret.org` is hosted by Zulip Cloud, since one could test URLs for presence in `REALM_HOSTS`, but I can't imagine why that'd be a useful security goal)

Reply:
Hmm, that does make sense, although I'm not completely convinced it's worth the complexity in this case (especially given that we've already chosen to have one violation of this invariant).

I'm not totally sure I understand the design of Camo — is it the case that if there's a bug in the logic to decide what's a user-uploaded image, it'll fail in a secure way? That is, the image will not show for anyone, since Camo doesn't have the required authorization? That would make sense, but i do just want to verify that we're not going to have another caching authenticated responses we don't expect to issue :)

Looking at the way we host an org that I know we do the domain name white-labeling for, we're doing it with a `CNAME`, which makes it quite obvious what's going on, and I think even if we didn't, the IP addresses all being the same would be a pretty big giveaway, I don't really think it's realistic to expect that we would want or be able to hide that.

@andersk, curious if you have thoughts on this — I'd like to decide how we want to do this before making more changes on this PR :)

Reply:
> I'm not totally sure I understand the design of Camo — is it the case that if there's a bug in the logic to decide what's a user-uploaded image, it'll fail in a secure way? That is, the image will not show for anyone, since Camo doesn't have the required authorization? That would make sense, but i do just want to verify that we're not going to have another caching authenticated responses we don't expect to issue :)

Yes, that's correct.

Reply:
It’s already easy to test a hostname for presence in `REALM_HOSTS` simply by issuing a request to determine whether Django considers it to be in `ALLOWED_HOSTS`.

However, this does bring up another issue: since the owner of chat\.secret\.org controls their own DNS, they could insert a proxy between chat\.secret\.org viewers and our server at any time. It’s of course expected that users of that realm have to trust that owner. But if we whitelist `REALM_HOSTS` here, we fail to prevent that owner from breaking our privacy goals on *other realms*.

So I think Tim’s inclination is right, and we should limit the whitelist to `realm.uri` (`= EXTERNAL_URI_SCHEME + realm.host`) and the domain of `STATIC_URL`:

https://github.com/zulip/zulip/blob/a5d2cadb33b3e6f77a6a52dffef11aa0b327af39/zproject/computed_settings.py#L548-L551

(which we can factor out as an additional variable there).